### PR TITLE
Enable HTTP/2 in SocketsHttpHandler via HttpRequestMessage.Version

### PR DESCRIPTION
**Pull request by stephentoub**
We now respect the version: if it's 1.1, we use 1.1, and if it's 2.0, we try to use HTTP/2.  At the same time, the default for HttpRequestMessage.Version goes back to 1.1, so to use HTTP/2, you opt-in by setting the version explicitly to 2.0.

This is almost all tweaking tests.  In a few cases, I noticed some other issues in the tests as I was doing so and fixed them / moved a few things around.

Fixes https://github.com/dotnet/corefx/issues/35138
Fixes https://github.com/dotnet/corefx/issues/37224
cc: @geoffkizer, @davidsh, @wfurt

### CODE REVIEW

File: src/System.Net.Http/src/ILLinkTrim.xml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 <linker>
   <assembly fullname="System.Net.Http">
+    <type fullname="System.Net.Http.HttpClient">
```
Comment (davidsh) [inline src/System.Net.Http/src/ILLinkTrim.xml]: What does this code do?  Just wondering...

Comment (stephentoub) [reply src/System.Net.Http/src/ILLinkTrim.xml]: There's a linker tool that runs as a post-compile step and tree shakes, removing unused internals and privates. The tests rely via reflection on this property's setter not being removed, so this entry in the config file read by the tool essentially keeps the property rooted and safe from removal.

**Comment by stephentoub** [comment]
Sure. I checked the grpc code before as well to validate it was setting the version.

File: src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs [hunk 2]
```diff
@@ -477,13 +477,192 @@ public sealed class SocketsHttpHandler_HttpClientHandler_Proxy_Test : HttpClient
...
 
-    public sealed class SocketsHttpHandler_HttpClientHandler_TrailingHeaders_Test : HttpClientHandlerTest_TrailingHeaders_Test
+    public class SocketsHttpHandler_HttpClientHandler_TrailingHeaders_Test : HttpClientHandlerTest_TrailingHeaders_Test
     {
         public SocketsHttpHandler_HttpClientHandler_TrailingHeaders_Test(ITestOutputHelper output) : base(output) { }
-        // PlatformHandlers don't support trailers.
         protected override bool UseSocketsHttpHandler => true;
     }
 
+    public sealed class SocketsHttpHandler_Http2_HttpClientHandler_TrailingHeaders_Test : HttpClientHandlerTestBase
```
Comment (wfurt) [inline src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: Why did you move Http2 trailing header tests here? And not http1? They should be pretty similar.

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: Because these were specific to SocketsHttpHandler and specific to HTTP/2.  Leaving them on the base class then had them be run and then skipped for all other handlers / configurations.  Seemed they should be where they actually target.

Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: I still don't understand why protocol version would matter. HttpClientHandlerTest_TrailingHeaders_Test  is only used by SocketsHttpHandlerTest

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: We don't support trailing headers on any other handler?  Then yeah, the HTTP/1.1 variants should be moved to this file as well.

Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: I don't think so since the property did not exist before. We made @caesar-chen to put tests for trailing headers to new file.

Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs]: > to put tests for trailing headers to new file.

A separate file is fine.  But these tests are specific to SocketsHttpHandler.  They're not for HttpClientHandler.
